### PR TITLE
Fix padding for single txn blocks

### PR DESCRIPTION
**Pull request by Nashtare**
We were still prepending to the actual txn the dummy payload for blocks containing only 1 txn, but we were then updating that payload if the block contained withdrawals.

We now do a case disjunction for single_txn blocks based on the presence of withdrawals:
- no withdrawals: dummy payload prepended, for efficiency
- withdrawals: dummy payload appended, to be then properly updated with withdrawals and modified state trie root

### CODE REVIEW

**Review by BGluth — APPROVED** [review]
Yeah looks good! This is something that I would not have caught on my own, so thanks for the PR.

File: protocol_decoder/src/decoding.rs [hunk 1]
```diff
@@ -317,9 +322,23 @@ impl ProcessedBlockTrace {
...
-                // Just need one.
-                let dummy_txn = create_dummy_gen_input(other_data, extra_data, initial_tries);
-                gen_inputs.insert(0, dummy_txn);
+                // We just need one dummy entry.
+                // If there are withdrawals, we will need to append them at the end of the block
+                // execution, in which case we directly append the dummy proof
+                // after the only txn of this block.
+                // If there are no withdrawals, then the dummy proof will be prepended to the
+                // actual txn.
+                match has_withdrawals {
```
Comment (dvdplm) [inline protocol_decoder/src/decoding.rs:331]: Why not use `if { … } else { … }`?

**Review by dvdplm — APPROVED** [review]

File: protocol_decoder/src/decoding.rs (hunk 1 above)
Comment (BGluth) [reply protocol_decoder/src/decoding.rs:331]: I actually find that using a `match` on a `bool` is usually more readable (there might be something wrong with me, idk :upside_down_face:) than an `if / else`. I think using a match is a bit less common though.

**Review by muursh — APPROVED** [review]

File: protocol_decoder/src/decoding.rs (hunk 1 above)
Comment (Nashtare) [reply protocol_decoder/src/decoding.rs:331]: Yeah not sure why I went for match, maybe an implicit preference as Brendan said, happy to change though

Comment (muursh) [reply protocol_decoder/src/decoding.rs:331]: I'd vote for keeping match

**Pull request merged**